File: services/web-api/src/main/java/com/github/jeffreyjpz/timestamped_for_f1_web_api/web/v1/drivers/dtos/Driver.java
package com.github.jeffreyjpz.timestamped_for_f1_web_api.web.v1.drivers.dtos;

import com.fasterxml.jackson.annotation.JsonAlias;
import com.fasterxml.jackson.annotation.JsonProperty;

import lombok.Getter;
import lombok.RequiredArgsConstructor;

@Getter
@RequiredArgsConstructor
public class Driver {

    @JsonAlias({"broadcastName", "broadcast_name"})
    @JsonProperty("broadcast_name")
    private final String broadcastName;

    @JsonAlias({"countryCode", "country_code"})
    @JsonProperty("country_code")
    private final String countryCode;

    @JsonAlias({"driverNumber", "driver_number"})
    @JsonProperty("driver_number")
    private final Integer driverNumber;

    @JsonAlias({"firstName", "first_name"})
    @JsonProperty("first_name")
    private final String firstName;

    @JsonAlias({"fullName", "full_name"})
    @JsonProperty("full_name")
    private final String fullName;

    @JsonAlias({"headshotUrl", "headshot_url"})
    @JsonProperty("headshot_url")
    private final String headshotUrl;

    @JsonAlias({"lastName", "last_name"})
    @JsonProperty("last_name")
    private final String lastName;

    @JsonAlias({"meetingKey", "meeting_key"})
    @JsonProperty("meeting_key")
    private final Integer meetingKey;

    @JsonAlias({"nameAcronym", "name_acronym"})
    @JsonProperty("name_acronym")
    private final String nameAcronym;

    @JsonAlias({"sessionKey", "session_key"})
    @JsonProperty("session_key")
    private final Integer sessionKey;

    @JsonAlias({"teamColour", "team_colour"})
    @JsonProperty("team_colour")
    private final String teamColour;

    @JsonAlias({"teamName", "team_name"})
    @JsonProperty("team_name")
    private final String teamName;

}

File: dashboard/src/types/driver.ts
import z from "zod";

export const driver = z
    .object({
        broadcast_name: z.string(),
        country_code: z.string().length(3),
        driver_number: z.number().int().positive(),
        first_name: z.string(),
        full_name: z.string(),
        headshot_url: z.url().nullable(),
        last_name: z.string(),
        meeting_key: z.number().int().positive(),
        name_acronym: z.string().max(3),
        session_key: z.number().int().positive(),
        team_colour: z.string().regex(/^[A-Fa-f0-9]{6}$/), // Hex colour
        team_name: z.string(),
    });

export const drivers = z.array(driver);

export const driverRead = driver
    .pick({
        driver_number: true,
        meeting_key: true,
        session_key: true
    });

export const driversRead = driver
    .extend({
        broadcast_name: z.array(z.string()),
        country_code: z.array(z.string().length(3)),
        driver_number: z.array(z.number().int().positive()),
        first_name: z.array(z.string()),
        full_name: z.array(z.string()),
        headshot_url: z.array(z.url()),
        last_name: z.array(z.string()),
        meeting_key: z.array(z.number().int().positive()),
        name_acronym: z.array(z.string().max(3)),
        session_key: z.array(z.number().int().positive()),
        team_colour: z.array(z.string().regex(/^[A-Fa-f0-9]{6}$/)),
        team_name: z.array(z.string()),
    })
    .partial();

export type Driver = z.infer<typeof driver>;
export type Drivers = z.infer<typeof drivers>;
export type DriverRead = z.infer<typeof driverRead>;
export type DriversRead = z.infer<typeof driversRead>;